src/csharp/GRPC-TOOLS-MSBUILD.md

```diff
@@ -0,0 +1,215 @@
+A new list of items - _Protobuf_Rooted_ - is created from the _Protobuf_ items with _ProtoRoot_ metadata set
+
+* Either already explicitly set in the item
+* Or “.” if in project directory
+* Or relative path to project directory
+
+Now prune out from _Protobuf_Rooted_ the items that the user doesn’t want to compile - those don’t have
+_ProtoCompile_ metadata as _‘true’_.  The pruned list is now called _Protobuf_Compile_.
+
+Set the _Source_ metadata on _Protobuf_Compile_ items to be the name of the ._proto_ file.
+The _Source_ metadata is used later as a key to map generated files to ._proto_ files.
+
+### Handling incremental builds
+
+#### Gathering files to check for incremental builds
+
+The target _Protobuf_PrepareCompile_ tries to work out which files the protobuf compiler will
+generate without actually calling the protobuf compiler. This is a best-effort guess.
+The custom task _ProtoCompilerOutputs_ is called to do this. The results are stored in the
+item list _Protobuf_ExpectedOutputs_.
+
+The target _Protobuf_PrepareCompile_ also reads previously written ._protodep_ files to get
+any actual files previously generated. The custom task _ProtoReadDependencies_ is called to
+do this. The results are stored in the item list _Protobuf_Dependencies_.
+This is in case the list of actual files is different from the previous best-effort guess
+from _ProtoCompilerOutputs_.
+
+The expected outputs and previous outputs are needed so that the timestamps of those files
+can be checked later when handling an incremental build.
+
+#### Understanding incremental builds
+
+To avoid unnecessarily recompiling the ._proto_ files during an incremental build the
+target _\_Protobuf_GatherStaleBatched_ tries to work out if any files have changed.
+
+It checks for out of date files using MSBuilds incremental build feature that compares the 
+timestamps on a target’s _Input_ files to its _Output_ files.
+See [How to: Build incrementally](https://learn.microsoft.com/en-us/visualstudio/msbuild/how-to-build-incrementally)
+
+The _Inputs_ that are checked are:
+* Timestamps of the ._proto_ files
+* Timestamps of previously generated files (list of these files read from ._protodep_ files)
+* Timestamps of MSBuild project files
+
+These are checked against the _Outputs_:
+* Timestamps of the expected generated files
+
+[MSBuild target batching](https://learn.microsoft.com/en-us/visualstudio/msbuild/item-metadata-in-target-batching)
+is used to check each ._proto_ file against its expected outputs. The batching is done by specifying the _Source_ 
+metadata in the _Input_. Items where _Source_ metadata matches in both input and output are in each batch.
+
+The target _\_Protobuf_GatherStaleBatched_ sets the metadata _\_Exec=true_ on _\_Protobuf_OutOfDateProto_ 
+items that are out of date.
+
+Later in the target _\_Protobuf_GatherStaleFiles_, the items in _\_Protobuf_OutOfDateProto_ that don’t have
```

Comment:
use backticks here and elsewhere, otherwise it's going to be very easy with the wrong property name being rendered. E.g. `_Protobuf_GatherStaleFiles`